lib/philomena/images.ex

```diff
@@ -427,6 +431,44 @@ defmodule Philomena.Images do
         {:ok, count}
     end)
     |> Repo.transaction()
+    |> case do
+      {:ok, %{image: {image, _added, _removed}}} = res ->
+        update_tag_change_limits_after_commit(image, attribution)
+
+        res
+
+      err ->
+        err
+    end
+  end
+
+  defp check_tag_change_limits_before_commit(image, attribution) do
+    user = attribution[:user]
+    ip = attribution[:ip]
+
+    cond do
+      Limits.is_limited_for_tag_count?(user, ip) ->
+        {:error, "Limit exceeded"}
+
+      Limits.is_limited_for_rating_count?(user, ip) ->
+        {:error, "Limit exceeded"}
+
+      is_nil(user) and length(image.added_tags) + length(image.removed_tags) > 100 ->
+        {:error, "Limit exceeded"}
+
+      true ->
+        {:ok, 0}
+    end
+  end
+
+  def update_tag_change_limits_after_commit(image, attribution) do
+    tag_changed_count = length(image.added_tags) + length(image.removed_tags)
+    rating_changed_count = image.ratings_changed
```

Comment:
This uses global `ratings_change` count for the image, meaning that if someone else updates the rating tags on that image (not necessarily the current anon), the current anon will be rate-limited on rating tag changes too even if they never changed any rating tags. This needs to use some kind of scoped/constrained counter. My suggestion is to scope the redis counters further.

Reply:
`ratings_changed` is a non-persisted field that is only ever assigned by the process of submitting a single change batch. Are you sure?

Reply:
derp, you're right, i was convinced it was persistent for some reason